pep-0677.rst

```diff
@@ -635,24 +635,92 @@ callable types and ``=>`` for lambdas.
-possible alternatives to the current plan.
+Evaluation and Structured API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We intend to create new builtin types to which the new AST nodes will
+evaluate, exposing them in the ``types`` module.
+
+The ``CallableType`` and ``AsyncCallableType`` AST nodes should evaluate
+to ``types.CallableType`` and ``types.AsyncCallableType``. Each of these
+should have fields:
+
+  - ``args`` containing an evaluated arguments list (see below)
+  - ``returns`` containing the evaluated return type
+
+The arguments list AST nodes should evaluate as follows:
+- ``AnyArguments`` evaluates to ``types.CallableAnyArguments``
+
+  - This type has no fields
+
+- ``ArgumentsList`` evaluates to ``types.CallableARgumentsList``, having
+
+  - A field ``posonlyargs`` which will be a tuple of all parameter types.
+
+- ``Concatenation`` evaluates to ``types.CallableConcatenation``, having
+
+  - A field ``posonlyargs`` which will be a tuple of the normal parameter types.
+  - A field ``param_spec`` which will be the final ``ParamSpec`` argument.
+
+Backward-Compatible API
+~~~~~~~~~~~~~~~~~~~~~~~
+
+To get backward compatibility with the existing ``types.Callable`` API,
+which relies on fields ``__args__`` and ``__parameters__``, we define them
+as follows:
+
+- For ``types.CallableType``:
+
+  - ``__args__`` should be the tuple result of appending the return type to the
+    result of fetching ``__args__`` on ``args`` (see below).
+  - ``__parameters__`` should be the result of concatenating ``__parameters__``
+    of the arguments list with ``__parameters__`` of the return type.
+
+- ``types.AsyncCallableType`` should work the same, except that its return
+  value should automatically be wrapped in a ``typing.Awaitable`` when
+  evaluating ``__args__``.
+- For ``types.CallableAnyAguments``:
+
+  - ``__args__`` should evaluate to ``(Ellipses,)``.
+  - ``__parameters__`` should evaluate to ``()``.
+
+- For ``types.CallableArgumentsList``:
+
+  - ``__args__`` should evaluate to ``posonlyargs``.
+  - ``__parameters__`` should be the result of concatenating the
+    ``__parameters__`` of each type in ``posonlyargs``.
+
+- For ``types.CallableConcatenation``:
+
+  - ``__args__`` should be the tuple result of appending ``param_spec`` to
+    ``posonlyargs``.
+  - ``__parameters__`` should be the result of concatinating the
+    ``__parameters__`` of each type in ``posonlyargs``, and also.
+
+
+
+Behavior of other methods
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- The ``__eq__`` method should treat equivalent ``typing.Callable``
+  values as equal to values constructed using the builtin syntax.
+- The ``__repr__`` method of ``types.CallableType`` and
+  ``types.AsyncCallableType`` should produce a representation in arrow syntax.
+
+  - The argument list types should produce a placeholder for ``__repr__``
+    rather than a pretty tuple-like value because they cannot be evaluated
+    except inside of a full arrow type literal.
```

Comment:
Not sure what you mean by "placeholder". Are you saying that for e.g. `(int, int) -> str` the argument list should look like `"(int, int)"` rather than `repr((int, int))` (which is `"(<class 'int'>, <class 'int'>)"`)?